scripts/bert/conversion_tools/convert_tf_model.py

```diff
@@ -19,34 +19,49 @@
 # pylint:disable=redefined-outer-name,logging-format-interpolation
 """ Script for converting TF Model to Gluon. """
 
-import os
-import logging
 import argparse
+import json
+import logging
+import os
+import sys
+
 import mxnet as mx
+
 from gluonnlp.model import BERTEncoder, BERTModel
 from gluonnlp.model.bert import bert_hparams
-from utils import convert_vocab, get_hash, read_tf_checkpoint
-
-parser = argparse.ArgumentParser(description='Conversion script for Tensorflow BERT model')
-parser.add_argument('--model', type=str, default='bert_12_768_12',
-                    help='BERT model name. options are bert_12_768_12 and bert_24_1024_16.'
-                         'Default is bert_12_768_12')
-parser.add_argument('--tf_checkpoint_dir', type=str,
-                    default=os.path.join('~', 'cased_L-12_H-768_A-12/'),
-                    help='Path to Tensorflow checkpoint folder. '
-                         'Default is /home/ubuntu/cased_L-12_H-768_A-12/')
-parser.add_argument('--out_dir', type=str,
+
+sys.path.insert(0, os.path.abspath(os.path.join(__file__, os.pardir, os.pardir)))
+
+from utils import (get_hash, load_tf_vocab, read_tf_checkpoint,
+                   tf_vocab_to_gluon_vocab)
+
+
+parser = argparse.ArgumentParser(
+    description='Conversion script for Tensorflow BERT model',
+    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+parser.add_argument('--model',
+                    type=str,
+                    default='bert_12_768_12',
+                    choices=['bert_12_768_12', 'bert_24_1024_16'],
+                    help='BERT model name')
+parser.add_argument('--tf_checkpoint_dir',
+                    type=str,
+                    help='Path to Tensorflow checkpoint folder.')
+parser.add_argument('--tf_config_name', type=str,
+                    default='bert_config.json',
+                    help='Name of Bert config file')
+parser.add_argument('--out_dir',
+                    type=str,
                     default=os.path.join('~', 'output'),
-                    help='Path to output folder. The folder must exist. '
-                         'Default is /home/ubuntu/output/')
+                    help='Path to output folder. The folder must exist.')
 parser.add_argument('--debug', action='store_true', help='debugging mode')
 args = parser.parse_args()
 logging.getLogger().setLevel(logging.DEBUG if args.debug else logging.INFO)
 logging.info(args)
 
 # convert vocabulary
 vocab_path = os.path.join(args.tf_checkpoint_dir, 'vocab.txt')
-vocab, reserved_token_idx_map = convert_vocab(vocab_path)
+vocab = tf_vocab_to_gluon_vocab(load_tf_vocab(vocab_path))
```

Comment:
This vocab is not compatible from the embedding of existing converted pre-trained models hosted on s3, right? Shall we use this only for new vocab like scibert/biobert?

Reply:
What do you mean with compatible? Loading the serialized vocabs on S3 that were created with the previous code is supported, and they will be loaded correctly.
Running the `convert_tf_model.py` will not result in an identical Vocabulary though. In particular, unlike before, the sorting of the vocabulary index as given in the tensorflow vocab is preserved. But that shouldn't be necessary: The conversion output (vocab + parameters) are used together and we don't need to mix an old parameter with a new vocabulary.

Would there be any reason for re-generating the exact json as is currently on S3? If not, we should just remove the old code (as done in this PR) as it is 50 lines, compared to 5 lines now. Also the old code is preserved in the git history.

Reply:
I was thinking about the use case where we want to use the checkpoint and vocab previously distributed on s3 and validate them again with TF model. This is useful if we want to check if a particular change in mxnet causes difference in the model precision (e.g. layernorm or multi-head attention implementations are changed or optimized in mxnet). If so, we cannot use `tf_vocab_to_gluon_vocab ` with the existing checkpoint. 

On second thought, we don't necessarily need to rely on this script if we want to check if any operator change in mxnet causes differences. We can run BERT QA inference with a new mxnet commit to check if accuracy/F1 matches. 

For this PR, would you mind adding some documentations here that the vocabulary of this script is different from the one we distributed in gluonnlp 0.6?

Reply:
But we can still use checkpoint and vocab that are previously distributed. The vocabulary used by tensorflow and the vocabulary used by gluon are independent. Or what do you mean?

I'm unsure what kind of documentation would be helpful here. The ordering of the vocabularies has never been documented / guaranteed, nor do people rely on it (the API is based on using the token_to_idx mapping, not based on hard coded indices). What use-case do you have in mind?

Reply:
If I use this script with `--gluon_dataset book_corpus_wiki_en_uncased --gluon_model bert_12_768_12 ` and leave `--gluon_parameter_file` as None, it would download the model from s3, generate a new vocab that has different indices from the one we host on s3, and the consistency check will fail, is that right? 

The embedding in the previous checkpoints assumes vocab[unk] = 0

Reply:
Thanks @eric-haibin-lin for the elaboration.

If `--gluon_parameter_file` is not specified, the vocabulary from S3 should be used. But you're right that this was not correctly implemented in the comparison script. It is addressed in https://github.com/dmlc/gluon-nlp/pull/735/commits/59b225060503519ba9acef072a7a38380322bfdd

I added a `--gluon_vocab_file` argument, that is mandatory to specify if `--gluon_parameter_file` is given. It expects users to specify the vocab file generated by the `convert_tf_model.py` script they used to generate the param file. If no  `--gluon_parameter_file` is specified and the parameters are loaded from S3, the vocab will also be loaded from S3.